### PR TITLE
Make sure ENV in the config is not cleared by setting `-e`

### LINKED ISSUE #560: Specifying env vars with 'fly deploy -e' wipes out env vars set in fly.toml
Setting env vars with `-e` should merge with those set on `fly.toml`, overriding only those with the same keys. Now, all env vars will be wiped out and replaced.

### PR DESCRIPTION
**Pull request by sudhirj**
Made ENV vars a property of the AppConfig, and made sure SetEnvVariable and SetEnvVariables don't clear out the entire set of vars.

Fixes #560

### CODE REVIEW

File: flyctl/app_config.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ type AppConfig struct {
 	AppName    string
 	Build      *Build
 	Definition map[string]interface{}
+	Env        map[string]string
```
Comment (azazeal) [inline flyctl/app_config.go:30]: I think there might be benefit to declaring a new type, `Environment`, on which the rest of the functionality of this PR should be implemented on. `Build` works similarly.

File: flyctl/app_config.go [hunk 2]
```diff
@@ -54,9 +56,7 @@ func LoadAppConfig(configFile string) (*AppConfig, error) {
 		return nil, err
 	}
 
-	appConfig := AppConfig{
-		Definition: map[string]interface{}{},
-	}
+	appConfig := *NewAppConfig()
```
Comment (azazeal) [inline flyctl/app_config.go:59]: Why are we dereferencing here?

File: flyctl/app_config_test.go [hunk 3]
```diff
@@ -54,3 +54,21 @@ func TestLoadTOMLAppConfigWithServices(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, p.Definition, rawData)
 }
+
+func TestLoadTOMLAppConfigWithEnvVars(t *testing.T) {
+	path := "./testdata/env-vars.toml"
+	p, err := LoadAppConfig(path)
+	assert.NoError(t, err)
```
Comment (azazeal) [inline flyctl/app_config_test.go:61]: This should probably be `require.NoError(t, err)` as the rest of the test is pointless unless the configuration is loaded.

File: flyctl/app_config_test.go [hunk 4]
```diff
@@ -54,3 +54,21 @@ func TestLoadTOMLAppConfigWithServices(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, p.Definition, rawData)
 }
+
+func TestLoadTOMLAppConfigWithEnvVars(t *testing.T) {
+	path := "./testdata/env-vars.toml"
```
Comment (azazeal) [inline flyctl/app_config_test.go:59]: Maybe this should be a constant `const path = "testdata/env-vars.toml"`.

File: flyctl/app_config.go (hunk 1 above)
Comment (sudhirj) [reply flyctl/app_config.go:30]: That sounds good.

File: flyctl/app_config.go (hunk 2 above)
Comment (sudhirj) [reply flyctl/app_config.go:59]: The method returns a pointer by default, which we actually don't want in this context. We should probably get the method to return a struct by default and let the caller take the address if they want.

File: flyctl/app_config_test.go (hunk 3 above)
Comment (sudhirj) [reply flyctl/app_config_test.go:61]: Oh, does assert not fail immediately? Will change.

File: flyctl/app_config.go (hunk 2 above)
Comment (azazeal) [reply flyctl/app_config.go:59]: I think this is part of [`func LoadAppConfig(configFile string) (*AppConfig, error) {`](https://github.com/superfly/flyctl/blob/89efffb440a3512f80944c30cb1832ff82f0acb7/flyctl/app_config.go#L53) which returns a pointer, no? Sorry if I'm missing something obvious!

Comment (azazeal) [reply flyctl/app_config.go:59]: Hi there again - let me clarify: I understand that in this particular context we used to do the following:

```
appConfig := AppConfig{
	Definition: map[string]interface{}{},
}
```

Changing from this to `appConfig := NewAppConfig()` (with `appConfig` ending up being a pointer) would not affect the rest of the code.

Sorry for the confusion!

Comment (sudhirj) [reply flyctl/app_config.go:60]: The last line (the return) was taking the address again – I've removed that now and made this a simple call to `NewAppConfig()`.

File: flyctl/app_config_test.go (hunk 4 above)
Comment (sudhirj) [reply flyctl/app_config_test.go:59]: Do you think I should standardise all the rest of the test as well? seems like a good idea.

File: flyctl/app_config.go (hunk 1 above)
Comment (sudhirj) [reply flyctl/app_config.go:30]: Actually that seems like it'll add more code for not much benefit — will need to add a `Get`, `Set` and a `Range` (all three are used) methods, or just expose the map again on the `Env` struct. Checking if see if there's any other benefit to pulling this up, or if we can just use one of the stdlib maps like Values.

Comment (sudhirj) [reply flyctl/app_config.go]: Sorry, misunderstood. I've added a new type called `Environment` which is just `map[string]string`.

**Comment by jsierles** [comment]
How do we feel about merging?

**Comment by rugwirobaker** [comment]
LGTM!!

**Comment by diwu1989** [comment]
Guys, does this really work?
I just upgraded my flyctl today and did a deploy, and all env variables in fly.toml got wiped out!

**Comment by michaeldwan** [comment]
@diwu1989 there's an issue, we'll revert and cut a new release

**Pull request merged**